press/press/doctype/deploy_candidate_build/deploy_candidate_build.py

```diff
@@ -0,0 +1,1167 @@
+		self.cleanup_build_directory()
+		self.build_steps.clear()
+		self.build_directory = None
+		self.build_error = ""
+		self.build_output = ""
+		# Failure flags
+		self.user_addressable_failure = False
+		self.manually_failed = False
+		# Build times
+		self.build_start = None
+		self.build_end = None
+		self.build_duration = None
+		# Pending times
+		self.pending_start = None
+		self.pending_end = None
+		self.pending_duration = None
+
+	def set_build_server(self):
+		if not self.build_server:
+			self.build_server = get_build_server(self.group)
+
+		if self.build_server or self.no_build:
+			return
+
+		throw_no_build_server()
+
+	def validate_status(self):
+		if self.status in ["Draft", "Success", "Failure", "Scheduled"]:
+			return True
+
+		frappe.throw(
+			f"Build is in <b>{self.status}</b> state. "
+			"Please wait for build to succeed or fail before retrying."
+		)
+		return False
+
+	def pre_build(self, **kwargs):
+		self.reset_build_state()
+		self.add_pre_build_steps()
+
+		if not self.validate_status():
+			return
+
+		if "no_cache" in kwargs:
+			self.no_cache = kwargs.get("no_cache")
+			del kwargs["no_cache"]
+
+		self.set_build_server()
+		self.set_status(Status.PENDING, timestamp_field="pending_start")
+
+		self.save()
+
+		(
+			user,
+			session_data,
+			team,
+		) = (
+			frappe.session.user,
+			frappe.session.data,
+			get_current_team(True),
+		)
+
+		frappe.set_user(frappe.get_value("Team", team.name, "user"))
+		queue = "default" if frappe.conf.developer_mode else "build"
+
+		frappe.enqueue_doc(
+			self.doctype,
+			self.name,
+			"_build",
+			queue=queue,
+			timeout=2400,
+			enqueue_after_commit=True,
+		)
+
+		frappe.set_user(user)
+		frappe.session.data = session_data
+		frappe.db.commit()
+
+	def on_update(self):
+		if self.status == "Running":
+			frappe.publish_realtime(
+				f"bench_deploy:{self.name}:steps",
+				doctype=self.doctype,
+				docname=self.name,
+				message={"steps": self.build_steps, "name": self.name},
+			)
+		else:
+			frappe.publish_realtime(
+				f"bench_deploy:{self.name}:finished",
+				doctype=self.doctype,
+				docname=self.name,
+			)
+
+		if self.has_value_changed("status") and self.candidate.team != "Administrator":
+			create_webhook_event("Bench Deploy Status Update", self, self.candidate.team)
+
+	def run_scheduled_build_and_deploy(self):
+		self.set_status(Status.DRAFT)
+		self.pre_build()
+
+	def after_insert(self):
+		if self.status != Status.SCHEDULED.value:
+			self.set_status(Status.DRAFT)
+			self.pre_build()
+
+	def autoname(self):
+		candidate_name = self.candidate.name[7:]
+		series = f"build-{candidate_name}-.######"
+		self.name = make_autoname(series)
+
+	def _stop_and_fail(self):
+		self.manually_failed = True
+		for job in get_background_jobs(self.doctype, self.name, status="started"):
+			if is_build_job(job):
+				stop_background_job(job)
+
+		self.set_status(Status.FAILURE)
+		self._set_build_duration()
+
+	def create_deploy(self):
```

Comment:
You can move this `deploy` functions to `Deploy Candidate` instead.

Reply:
There also exists a whitelisted `deploy` function in `build`, the idea was to deploy each build separately, therefore added basic deploy functions to `build`.

Reply:
That's fine then.